File: lab2/gr9sub1/ТиганинВА/src/fork_example.py
import sys
import os
import time

def child_process(name):
    print(f"{name}: PID = {os.getpid()}, PPID = {os.getppid()}", flush=True)
    sys.exit(0)

def main():
    print(f"Parent: My PID is {os.getpid()}", flush=True)

    fork_result1 = os.fork()
    if fork_result1 == 0:
        child_process("child_A")
    
    if fork_result1 > 0:
        fork_result2 = os.fork()
        if fork_result2 == 0:
            child_process("child_B")

    time.sleep(2)

    if fork_result1 > 0 and fork_result2 > 0:
        child_pid1, status1 = os.waitpid(fork_result1, 0)
        exit_code1 = os.WEXITSTATUS(status1)
        print(f"Parent: Child A (PID={child_pid1}) finished with status {exit_code1}", flush=True)

        child_pid2, status2 = os.waitpid(fork_result2, 0)
        exit_code2 = os.WEXITSTATUS(status2)
        print(f"Parent: Child B (PID={child_pid2}) finished with status {exit_code2}", flush=True)

        print("Parent: All children have finished", flush=True)

if __name__ == "__main__":
    main()

File: lab2/gr8sub1/Филипчик_Дима/src/fork_example.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/wait.h>

int main() {
    pid_t child_a, child_b;
    int status;

    printf("parent: started with PID=%d\n", getpid());
    fflush(stdout); 

    child_a = fork();
    if (child_a == 0) {
        printf("child_A: my PID=%d, parent PID=%d\n", getpid(), getppid());
        fflush(stdout);
        sleep(2);
        exit(0);
    }
	
    child_b = fork();
    if (child_b == 0) {
        printf("child_B: my PID=%d, parent PID=%d\n", getpid(), getppid());
        fflush(stdout);
        sleep(1);
        exit(0);
    }

    waitpid(child_a, &status, 0);
    waitpid(child_b, &status, 0);

    printf("parent: all children finished, exiting\n");
    return 0;
}
